### PR TITLE
Add friendly goal as an obstacle and fix robots getting stuck in the corner of enemy defense area

### LINKED ISSUE #2700: HRVO: Path plan around the goal
### Description of the task

HRVO doesn't path plan around the goal. This affects the Goalie and occasionally affects the ball placement robot in BallPlacementPlay.

If ENLSVG doesn't have a AVOID_FIELD_BOUNDARY_ZONE or a FRIENLDY/ENEMY_DEFENCE_AREA constraint, then ENLSVG may path plan through the goal and HRVO gets stuck and collides against the net.

1. Motion Constraint (Enum) which defines the list of motion constraints: https://github.com/UBC-Thunderbots/Software/blob/54d72ca024666aa920f28d71d44b7a00675f4469/src/proto/primitive.proto#L26-L48
2. Add the new motion constraint to the MotionConstraintSetBuilder: https://github.com/UBC-Thunderbots/Software/blob/424b79b84f40a4893411ea47131ec540efdcf1e6/src/software/ai/motion_constraint/motion_constraint_set_builder.cpp#L34-L80
3. Add the new enum to the switch statement and create an obstacle (some shape, probably a Polygon) which represents the new motion constraint enum you have added: https://github.com/UBC-Thunderbots/Software/blob/e6486a69c29f2273825b482b41eea616ebf661d0/src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp#L30-L115
4. Visually check in Thunderscope that the obstacle is drawn (obstacles are represented by orange lines)
5. Add a test to the set of path planner tests which we have in `software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp`

### Acceptance criteria

- [ ] Update `motion_constraint_visitor.cpp`
- [ ] Re-enable tests disabled due to this bug

### PR DESCRIPTION
**Pull request by nimazareian**
### Description
Add friendly goal as an obstacle and fix robots getting stuck in the corner of enemy defense area. 

### Testing Done
New tests added. In the process, some other edge case bugs were found.

### Resolved Issues
Resolves #2700
Quick fix for #2690 though not fully sure whether it will continue to happen or not as other edge case bugs were found in ENLSVG.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp [hunk 1]
```diff
@@ -684,6 +684,70 @@ TEST_F(TestEnlsvgPathPlanner, test_going_around_defense_area)
...
+    EXPECT_EQ(dest, path.value().back());
+
+    // Make sure path does not exceed a bounding box
+    Rectangle bounding_box({3, 3}, {4.5, -3});
+    TestUtil::checkPathDoesNotExceedBoundingBox(path.value(), bounding_box);
+    TestUtil::checkPathDoesNotIntersectObstacle(path.value(), obstacles);
+}
+
+// TODO (#2690): Path slightly intersects enemy defense area obstacle
+TEST_F(TestEnlsvgPathPlanner, DISABLED_test_going_from_below_enemy_defense_area_to_above)
```
Comment (itsarune) [inline src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp]: does it graze the inflated enemy defense area or the actual enemy defense area

**Review by itsarune — COMMENTED** [review]
looks good, minor question

File: src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp (hunk 1 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp]: In this test case, the path looks something like this, where it just slightly goes into the obstacle. Not really a huge deal imo, atleast compared to the other disabled test trying to draw the same path in reverse.
image

Comment (itsarune) [reply src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp]: Okay in this case, I suggest changing the validation so that it doesn't check if it intersects the obstacle ptr but the actually field enemy defence area itself. The whole point of the inflated defence area is to prevent these resolution discrepancies from the grid-based path planner

File: src/software/ai/navigator/path_planner/enlsvg_path_planner.cpp [hunk 2]
```diff
@@ -70,8 +70,8 @@ std::optional<Path> EnlsvgPathPlanner::findPath(const Point &start,
     EnlsvgPoint enlsvg_end   = convertPointToEnlsvgPoint(end);
 
     // closest unblocked points to requested start and end
-    auto new_start = findClosestUnblockedEnlsvgPoint(enlsvg_start);
-    auto new_end   = findClosestUnblockedEnlsvgPoint(enlsvg_end);
+    auto new_start = findClosestUnblockedEnlsvgPoint(enlsvg_start, enlsvg_end);
```
Comment (JoshuaWu7) [inline src/software/ai/navigator/path_planner/enlsvg_path_planner.cpp:73]: Have we considered if both `enlsvg_start` and `enlsvg_end` are blocked? Is this possible?

Comment (nimazareian) [reply src/software/ai/navigator/path_planner/enlsvg_path_planner.cpp:73]: We do check for that already. If either of the points are blocked and we cant find a nearby unblocked point to, then we just don't create a path.

https://github.com/UBC-Thunderbots/Software/blob/e27d6a8a7ec175a11deda516f5d8187f05fafcde/src/software/ai/navigator/path_planner/enlsvg_path_planner.cpp#L72-L83

Does that answer your question?

File: src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp (hunk 1 above)
Comment (nimazareian) [reply src/software/ai/navigator/path_planner/enlsvg_path_planner_test.cpp]: Sounds good. I've updated it.

**Review by JoshuaWu7 — DISMISSED** [review]
LGTM:+1:

File: src/software/geom/polygon.cpp
Comment (macallih) [inline src/software/geom/polygon.cpp:1]: ooh I like this!

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp [hunk 3]
```diff
@@ -112,6 +112,32 @@ RobotNavigationObstacleFactory::createStaticObstaclesFromMotionConstraint(
...
+            // Bottom goal post
+            obstacles.push_back(std::make_shared<GeomObstacle<Polygon>>(
+                Polygon::fromSegment(Segment(friendly_goal.posXNegYCorner(),
+                                             friendly_goal.negXNegYCorner()),
+                                     0, goal_obstacle_radius)));
+            // Left goal wall
+            obstacles.push_back(std::make_shared<GeomObstacle<Polygon>>(
+                Polygon::fromSegment(Segment(friendly_goal.negXPosYCorner(),
+                                             friendly_goal.negXNegYCorner()),
+                                     goal_obstacle_radius)));
```
Comment (macallih) [inline src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp:138]: how come you can omit the third argument here?

**Review by williamckha — DISMISSED** [review]

**Review by itsarune — DISMISSED** [review]
LGTM, pending CI

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp (hunk 3 above)
Comment (nimazareian) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp:138]: Good question. To avoid having to update every existing usecase of `Polygon::fromSegment` from having 1 radius argument to having 2 arguments, I decided to overload the function with two signatures, one takes 2 arguments and has the full implementation of the function
```cpp
    static Polygon fromSegment(const Segment& segment, double length_radius,
                               double width_radius);
```
and another which takes only 1 radius argument (similar to the one that was already in the code)
```cpp
    static Polygon fromSegment(const Segment& segment, double radius);
```
To avoid duplication of code, the method signature which takes 1 argument basically calls the signature that takes two arguments but with the same value for both `length_radius` and `width_radius`.
E.g. `fromSegment(Segment(), 2.5)` Calls `fromSegment(Segment(), 2.5, 2.5)`. C++ (and most other languages) is smart enough two differentiate the two implementations based on how many arguments you use to call the function.

**Review by williamckha — APPROVED** [review]

**Review by itsarune — APPROVED** [review]
LGTM!
